---
title: "Bisplotti User Guide"
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('bisplotti')`"
output:
    BiocStyle::html_document:
        highlight: pygments
        toc_float: true
        fig_width: 8
        git_height: 6
vignette: >
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteIndexEntry{Bisplotti User Guide}
    %\VignetteEncoding[utf8]{inputenc}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

# Bisplotti

`bisplotti` is a package to generate commonly produced plots in DNA methylation sequencing analyses. It creates plots
from standard Bioconductor formats (i.e., `GRanges`) and the commonly used `BSseq` format.

# Quick Start

## Installing

A development version is available on GitHub and can be installed via:
```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE)) {
    install.packages("BiocManager")
}

BiocManager::install("huishenlab/bisplotti")
```

```{r loadLib}
library(bisplotti)
```

# Create Plots

## Epiread

To create the lollipop epiread plots, two commands are needed. First, use the GRanges output from
`biscuiteer::readEpibed()` as input to `tabulateEpibed`, which turns the GRanges into a convenient matrix format. The
second command, `plotEpiread` uses the matrix output from `tabulateEpibed` as the input to create the lollipop plot.

```{r}
epibed.nome     <- system.file("extdata", "hct116.nome.epiread.gz", package="biscuiteer")
epibed.nome.gr  <- readEpibed(epibed = epibed.nome, is.nome = TRUE, genome = "hg19", chr = "chr1")
epibed.tab.nome <- tabulateEpibed(epibed.nome.gr)
plotEpiread(epibed.tab.nome$gc_table)
```

## Multiscale

The multiscale plot is based on Figures [2](https://www.nature.com/articles/s41588-018-0073-4/figures/2) and
[6](https://www.nature.com/articles/s41588-018-0073-4/figures/6) of
[Zhou et al., Nature Genetics, 2018](https://www.nature.com/articles/s41588-018-0073-4). It shows the average
methylation levels across varying size genomic windows. The example shows a small portion of chromosome 16 for window
sizes running from 1 Mb to 10 Mb.

```{r}
files.loc <- system.file("extdata", package="bisplotti")

files <- lapply(
    list.files(files.loc, pattern="Heyn_2012_100yr", full.names=TRUE), function(x) {
        return(rtracklayer::import(x, format="bedGraph"))
    }
)
cnames <- list.files(files.loc, pattern="Heyn_2012_100yr")
cnames <- gsub("Heyn_2012_100yr", "100yr", cnames)
cnames <- gsub(".bed.gz", "", cnames)
names(files) <- cnames

files.grl <- as(files, "GRangesList")

multiscaleMethylationPlot(files.grl)
```

## 1D Methylation Level Density

The 1D methylation level density plot shows the density of the methylation levels for all CpGs in your dataset. An
example of creating this plot is:
```{r}
bed <- system.file("extdata", "MCF7_Cunha_chr11p15.bed.gz", package="biscuiteer")
vcf <- system.file("extdata", "MCF7_Cunha_header_only.vcf.gz", package="biscuiteer")
bisc <- readBiscuit(BEDfile = bed, VCFfile = vcf, merged = FALSE)

meth1DDensity(bisc)
```
A matrix of methylation levels (beta values) can also be used as input to `meth1DDensity()`. This method assumes the
column names of your matrix are the samples in the matrix. The row names can either be NULL or the CpG loci/probes.

## 2D Methylation Level Density

The 2D methylation level density plot compares the density of average methylation levels in provided bins across two
samples. An example of creating this plot is:
```{r}
orig_bed <- system.file("extdata", "MCF7_Cunha_chr11p15.bed.gz", package="biscuiteer")
orig_vcf <- system.file("extdata", "MCF7_Cunha_header_only.vcf.gz", package="biscuiteer")
bisc1    <- readBiscuit(BEDfile = orig_bed, VCFfile = orig_vcf, merged = FALSE)

shuf_bed <- system.file("extdata", "MCF7_Cunha_chr11p15_shuffled.bed.gz", package="biscuiteer")
shuf_vcf <- system.file("extdata", "MCF7_Cunha_shuffled_header_only.vcf.gz", package="biscuiteer")
bisc2    <- readBiscuit(BEDfile = shuf_bed, VCFfile = shuf_vcf, merged = FALSE)

comb <- unionize(bisc1, bisc2)

meth2DDensity(comb, chr="chr11", sample_1 = "MCF7_Cunha", sample_2 = "MCF7_Cunha_shuffled")
```
A matrix of methylation levels (beta values) can also be used as input to `meth2DDensity()`. This method assumes the
column names of your matrix are the samples in the matrix. The row names must be the CpG loci/probes (i.e., the result
of `rownames(mat) <- as.character(granges(gr))`. In either input case, there must be at least two samples in your input
object.
